File: src/main/java/at/yawk/wm/wallpaper/animate/AnimationBuilder.java
package at.yawk.wm.wallpaper.animate;

import at.yawk.wm.Util;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import javax.annotation.Nullable;

/**
 * @author yawkat
 */
class AnimationBuilder {
    private static final Frame EMPTY_FRAME = new Frame();

    static {
        EMPTY_FRAME.setX(0);
        EMPTY_FRAME.setY(0);
        EMPTY_FRAME.setWidth(0);
        EMPTY_FRAME.setHeight(0);
        EMPTY_FRAME.setData(new byte[0]);
    }

    private static class ImageHolder {
        long time;
        BufferedImage image;

        @java.beans.ConstructorProperties({ "time", "image" })
        public ImageHolder(long time, BufferedImage image) {
            this.time = time;
            this.image = image;
        }

        public long getTime() {
            return this.time;
        }

        public BufferedImage getImage() {
            return this.image;
        }

        public boolean equals(Object o) {
            if (o == this) { return true; }
            if (!(o instanceof ImageHolder)) { return false; }
            final ImageHolder other = (ImageHolder) o;
            if (this.time != other.time) { return false; }
            final Object this$image = this.image;
            final Object other$image = other.image;
            if (this$image == null ? other$image != null : !this$image.equals(other$image)) { return false; }
            return true;
        }

        public int hashCode() {
            final int PRIME = 59;
            int result = 1;
            final long $time = this.time;
            result = result * PRIME + (int) ($time >>> 32 ^ $time);
            final Object $image = this.image;
            result = result * PRIME + ($image == null ? 0 : $image.hashCode());
            return result;
        }

        public String toString() {
            return "at.yawk.wm.wallpaper.animate.AnimationBuilder.ImageHolder(time=" + this.time + ", image=" +
                   this.image +
                   ")";
        }
    }

    /**
     * This method loads animations from a directory of the following format:
     *
     * <pre>
     * base.png
     * start100.png
     * start200.png
     * start300.png
     * [...]
     * stop100.png
     * stop200.png
     * stop300.png
     * [...]
     * </pre>
     *
     * The number represents the milliseconds when to show the frame. Every image must have the same size.
     */
    public static AnimatedWallpaper loadDirectory(Path animationDirectory) throws IOException {
        BufferedImage base = Util.INSTANCE.loadImage(animationDirectory.resolve("base.png"));

        AnimatedWallpaper wallpaper = new AnimatedWallpaper();
        wallpaper.setBaseFrame(createFrame(base, null, 0, 0, base.getWidth(), base.getHeight()));

        List<ImageHolder> start = new ArrayList<>();
        List<ImageHolder> stop = new ArrayList<>();

        Pattern framePattern = Pattern.compile("(start|stop)(\\d+)\\.png");

        Stream<Path> files = Files.list(animationDirectory);
        for (Path path : (Iterable<Path>) files::iterator) {
            Matcher matcher = framePattern.matcher(path.getFileName().toString());
            if (matcher.matches()) {
                long time = Long.parseUnsignedLong(matcher.group(2));
                ImageHolder holder = new ImageHolder(time, Util.INSTANCE.loadImage(path));
                (matcher.group(1).equals("start") ? start : stop).add(holder);
            }
        }

        wallpaper.setStart(linkAnimations(wallpaper.getBaseFrame(), start));
        wallpaper.setStop(linkAnimations(wallpaper.getBaseFrame(), stop));

        return wallpaper;
    }

    private static FrameAnimation linkAnimations(Frame background, List<ImageHolder> images) {
        FrameAnimation animation = new FrameAnimation();
        animation.setFrames(new ArrayList<>());

        // sort by time
        images.sort(Comparator.comparingLong(ImageHolder::getTime));

        for (ImageHolder image : images) {
            if (image.time == 0) { continue; }
            long interval = animation.getInterval();
            animation.setInterval(interval == 0 ? image.time : gcd(interval, image.getTime()));
        }

        Frame screen = new Frame(background);
        for (ImageHolder image : images) {
            Frame diff = diffFrame(screen, image.getImage());

            if (!diff.isEmpty()) {
                int i = (int) (image.time / animation.getInterval());
                // fill up with empty frames
                for (int j = animation.getFrames().size(); j <= i; j++) {
                    animation.getFrames().add(EMPTY_FRAME);
                }

                // add our frame
                animation.getFrames().set(i, diff);

                paint(screen, diff);
            }
        }

        return animation;
    }

    private static long gcd(long a, long b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    /**
     * Create a frame that represents the difference of the two input images. Pixel data will come from the front
     * image.
     */
    private static Frame diffFrame(Frame back, BufferedImage front) {
        int minX = back.getWidth();
        int minY = back.getHeight();
        int maxX = -1;
        int maxY = -1;
        for (int x = 0; x < back.getWidth(); x++) {
            for (int y = 0; y < back.getHeight(); y++) {
                int f = front.getRGB(x, y);
                if (f >>> 24 != 0) {
                    int bgOffset = (x + y * back.getWidth()) * 3;
                    boolean write = ((f >> 16) & 0xff) != (back.getData()[bgOffset] & 0xff);
                    write |= ((f >> 8) & 0xff) != (back.getData()[bgOffset + 1] & 0xff);
                    write |= (f & 0xff) != (back.getData()[bgOffset + 2] & 0xff);

                    if (write) {
                        minX = Math.min(minX, x);
                        minY = Math.min(minY, y);
                        maxX = Math.max(maxX, x);
                        maxY = Math.max(maxY, y);
                    }

                } // else {} foreground has 0 alpha, ignore pixel
            }
        }

        if (maxX < minX || maxY < minY) {
            // no pixels to copy
            return EMPTY_FRAME;
        } else {
            return createFrame(front, back, minX, minY, maxX - minX + 1, maxY - minY + 1);
        }
    }

    private static Frame createFrame(BufferedImage image, @Nullable Frame bg,
                                     int startX, int startY, int width, int height) {
        byte[] data = new byte[width * height * 3];
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                int rgba = image.getRGB(x + startX, y + startY);

                int a = rgba >>> 24 & 0xff;
                int r = rgba >>> 16 & 0xff;
                int g = rgba >>> 8 & 0xff;
                int b = rgba & 0xff;

                if (a != 0xff) {
                    double weightFront = a / 0xffD;
                    double weightBack = 1 - weightFront;

                    int backR = 0;
                    int backG = 0;
                    int backB = 0;
                    if (bg != null) {
                        int bgOffset = ((x + startX) + (y + startY) * bg.getWidth()) * 3;
                        backR = bg.getData()[bgOffset] & 0xff;
                        backG = bg.getData()[bgOffset + 1] & 0xff;
                        backB = bg.getData()[bgOffset + 2] & 0xff;
                    }

                    r = (int) (r * weightFront + backR * weightBack);
                    g = (int) (g * weightFront + backG * weightBack);
                    b = (int) (b * weightFront + backB * weightBack);
                }

                int offset = (x + y * width) * 3;
                data[offset] = (byte) r;
                data[offset + 1] = (byte) g;
                data[offset + 2] = (byte) b;
            }
        }
        Frame frame = new Frame();
        frame.setX(startX);
        frame.setY(startY);
        frame.setWidth(width);
        frame.setHeight(height);
        frame.setData(data);
        return frame;
    }

    /**
     * Paint the second frame onto the canvas frame, modifying the former.
     */
    private static void paint(Frame canvas, Frame frame) {
        for (int y = 0; y < frame.getHeight(); y++) {
            int frontOffset = (y * frame.getWidth()) * 3;
            int backOffset = ((frame.getX() - canvas.getX()) +
                              (y + frame.getY() - canvas.getY()) * canvas.getWidth()) * 3;
            System.arraycopy(frame.getData(), frontOffset, canvas.getData(), backOffset, 3 * frame.getWidth());
        }
    }
}

File: src/main/java/at/yawk/wm/wallpaper/animate/AnimationBuilder.kt
package at.yawk.wm.wallpaper.animate

import at.yawk.wm.Util
import java.awt.image.BufferedImage
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.regex.Pattern

private val EMPTY_FRAME = Frame(x = 0, y = 0, width = 0, height = 0, data = ByteArray(0))
/**
 * @author yawkat
 */
internal object AnimationBuilder {

    private data class ImageHolder(val time: Long, val image: BufferedImage)

    /**
     * This method loads animations from a directory of the following format:

     *
     * base.png
     * start100.png
     * start200.png
     * start300.png
     * [...]
     * stop100.png
     * stop200.png
     * stop300.png
     * [...]
     *

     * The number represents the milliseconds when to show the frame. Every image must have the same size.
     */
    fun loadDirectory(animationDirectory: Path): AnimatedWallpaper {
        val base = Util.loadImage(animationDirectory.resolve("base.png"))

        val baseFrame = createFrame(base, null, 0, 0, base.width, base.height)

        val start = ArrayList<ImageHolder>()
        val stop = ArrayList<ImageHolder>()

        val framePattern = Pattern.compile("(start|stop)(\\d+)\\.png")

        val files = Files.list(animationDirectory)
        for (path in Iterable { files.iterator() }) {
            val matcher = framePattern.matcher(path.fileName.toString())
            if (matcher.matches()) {
                val time = java.lang.Long.parseUnsignedLong(matcher.group(2))
                val holder = ImageHolder(time, Util.loadImage(path))
                (if (matcher.group(1) == "start") start else stop).add(holder)
            }
        }

        return AnimatedWallpaper(
                baseFrame = baseFrame,
                start = linkAnimations(baseFrame, start),
                stop = linkAnimations(baseFrame, stop)
        )
    }

    private fun linkAnimations(background: Frame, images: MutableList<ImageHolder>): FrameAnimation {
        images.sortBy { it.time }

        var interval: Long? = null
        for (image in images) {
            if (image.time == 0L) {
                continue
            }
            interval = if (interval == null) image.time else gcd(interval, image.time)
        }

        val frames = ArrayList<Frame>()
        val screen = background.copy()
        for (image in images) {
            val diff = diffFrame(screen, image.image)

            if (!diff.isEmpty()) {
                val i = (image.time / interval!!).toInt()
                // fill up with empty frames
                for (j in frames.size..i) {
                    frames.add(EMPTY_FRAME)
                }

                // add our frame
                frames[i] = diff

                paint(screen, diff)
            }
        }

        return FrameAnimation(interval!!, frames)
    }

    private fun gcd(a: Long, b: Long): Long {
        return if (b == 0L) a else gcd(b, a % b)
    }

    /**
     * Create a frame that represents the difference of the two input images. Pixel data will come from the front
     * image.
     */
    private fun diffFrame(back: Frame, front: BufferedImage): Frame {
        var minX = back.width
        var minY = back.height
        var maxX = -1
        var maxY = -1
        for (x in 0..back.width - 1) {
            for (y in 0..back.height - 1) {
                val f = front.getRGB(x, y)
                if (f.ushr(24) != 0) {
                    val bgOffset = (x + y * back.width) * 3
                    val write = (f shr 16 and 0xff != back.data[bgOffset].toInt() and 0xff) or
                            (f shr 8 and 0xff != back.data[bgOffset + 1].toInt() and 0xff) or
                            (f and 0xff != back.data[bgOffset + 2].toInt() and 0xff)

                    if (write) {
                        minX = Math.min(minX, x)
                        minY = Math.min(minY, y)
                        maxX = Math.max(maxX, x)
                        maxY = Math.max(maxY, y)
                    }

                } // else foreground has 0 alpha, ignore pixel
            }
        }

        if (maxX < minX || maxY < minY) {
            // no pixels to copy
            return EMPTY_FRAME
        } else {
            return createFrame(front, back, minX, minY, maxX - minX + 1, maxY - minY + 1)
        }
    }

    private fun createFrame(image: BufferedImage, bg: Frame?,
                            startX: Int, startY: Int, width: Int, height: Int): Frame {
        val data = ByteArray(width * height * 3)
        for (x in 0..width - 1) {
            for (y in 0..height - 1) {
                val rgba = image.getRGB(x + startX, y + startY)

                val a = (rgba ushr 24) and 0xff
                var r = (rgba ushr 16) and 0xff
                var g = (rgba ushr 8) and 0xff
                var b = rgba and 0xff

                if (a != 0xff) {
                    val weightFront = (a / 0xffD).toDouble()
                    val weightBack = 1 - weightFront

                    var backR = 0
                    var backG = 0
                    var backB = 0
                    if (bg != null) {
                        val bgOffset = (x + startX + (y + startY) * bg.width) * 3
                        backR = bg.data[bgOffset].toInt() and 0xff
                        backG = bg.data[bgOffset + 1].toInt() and 0xff
                        backB = bg.data[bgOffset + 2].toInt() and 0xff
                    }

                    r = (r * weightFront + backR * weightBack).toInt()
                    g = (g * weightFront + backG * weightBack).toInt()
                    b = (b * weightFront + backB * weightBack).toInt()
                }

                val offset = (x + y * width) * 3
                data[offset] = r.toByte()
                data[offset + 1] = g.toByte()
                data[offset + 2] = b.toByte()
            }
        }
        return Frame(
                startX, startY,
                width, height,
                data)
    }

    /**
     * Paint the second frame onto the canvas frame, modifying the former.
     */
    private fun paint(canvas: Frame, frame: Frame) {
        for (y in 0..frame.height - 1) {
            val frontOffset = y * frame.width * 3
            val backOffset = (frame.x - canvas.x + (y + frame.y - canvas.y) * canvas.width) * 3
            System.arraycopy(frame.data, frontOffset, canvas.data, backOffset, 3 * frame.width)
        }
    }
}
